src/java/org/apache/cassandra/cql3/statements/AlterTableStatement.java

```diff
@@ -346,9 +356,79 @@ public void validate(ClientState state)
+     * 3 things:
+     *   1) that all nodes are on 3.0+. We need this because 2.x nodes don't advertise their sstable versions.
+     *   2) for 3.0+, we use the new (CASSANDRA-15897) sstables versions set gossiped by all nodes to ensure all
+     *      sstables have been upgraded cluster-wise.
+     *   3) if the cluster still has some 3.0 nodes that predate CASSANDRA-15897, we will not have the sstable versions
+     *      for them. In that case, we also refuse DROP COMPACT (even though it may well be safe at this point) and ask
+     *      the user to upgrade all nodes.
+     */
+    private void validateCanDropCompactStorage()
+    {
+        Set<InetAddress> before3 = new HashSet<>();
+        Set<InetAddress> preC15897nodes = new HashSet<>();
+        Set<InetAddress> with2xSStables = new HashSet<>();
+        Splitter onComma = Splitter.on(',').omitEmptyStrings().trimResults();
+        for (InetAddress node : StorageService.instance.getTokenMetadata().getAllEndpoints())
+        {
+            if (MessagingService.instance().getVersion(node) < MessagingService.VERSION_30)
+            {
+                before3.add(node);
+                continue;
+            }
+
+            String sstableVersionsString = Gossiper.instance.getApplicationState(node, ApplicationState.SSTABLE_VERSIONS);
+            if (sstableVersionsString == null)
+            {
+                preC15897nodes.add(node);
+                continue;
+            }
+
+            try
+            {
+                // Note: 'storeRows' is basically a marker for 3.0+ sstables.
+                boolean has2xSStables = onComma.splitToList(sstableVersionsString)
+                                               .stream()
+                                               .map(VersionAndType::fromString)
+                                               .anyMatch(v -> !v.version().storeRows());
+                if (has2xSStables)
+                    with2xSStables.add(node);
+            }
+            catch (IllegalArgumentException e)
+            {
+                // Means VersionType::fromString didn't parse a version correctly. Which shouldn't happen, we shouldn't
+                // have garbage in Gossip. But crashing the request is not ideal, so we log the error but ignore the
+                // node otherwise.
+                noSpamLogger.error("Unexpected error parsing sstable versions from gossip for {} (gossiped value " +
+                                   "is '{}'). This is a bug and should be reported. Cannot ensure that {} has no " +
+                                   "non-upgraded 2.x sstables anymore. If after this DROP COMPACT STORAGE some old " +
+                                   "sstables cannot be read anymore, please use `upgradesstables` with the " +
+                                   "`--force-compact-storage-on` option.", node, sstableVersionsString, node);
```

Comment:
I was thinking of this one and I think we can leave it. The case is almost impossible, if it happens it would show a bug and we are giving guidance to deal with it.